File: Sources/Engine/session/SKKRegistration.cpp
//
//  SKKRegistration.cpp
//  AquaSKKEngine
//
//  Created by mzp on 2025/03/07.
//

#include "SKKRegistration.h"
#import <AquaSKKEngine/AquaSKKEngine-Preamble.h>
#import <AquaSKKEngine/AquaSKKEngine-Swift.h>

SKKRegistration::SKKRegistration()
:impl_(new SwiftObject(AquaSKKEngine::SKKRegistrationImpl::init())){

}
//    : state_(SKKRegistrationState::None) {}

void SKKRegistration::Start() {
//    state_ = SKKRegistrationState::Started;
    (*impl_)->start();
}

void SKKRegistration::Finish(const std::string &str) {
//    state_ = SKKRegistrationState::Finished;
//    word_ = str;
    (*impl_)->finish(str);
}

void SKKRegistration::Abort() {
//    state_ = SKKRegistrationState::Aborted;
//    word_.clear();
    (*impl_)->abort();
}

void SKKRegistration::Clear() {
//    state_ = SKKRegistrationState::None;
//    word_.clear();
    (*impl_)->clear();
}

const SKKRegistrationState SKKRegistration::getState() const {
//    return state_;
    (*impl_)->getState();
}
const std::string SKKRegistration::getWord() const SWIFT_COMPUTED_PROPERTY {
//    return word_;
    (*impl_)->getWord();
}

const std::string &SKKRegistration::Word() const {
//    return word_;
    (*impl_)->getWord();
}

File: Sources/Engine/session/SKKRegistration.swift
//
//  SKKRegistration.swift
//  AquaSKKEngine
//
//  Created by mzp on 2025/03/07.
//

import Foundation

public struct SKKRegistrationImpl {
    public var word: String
    public var state: SKKRegistrationState

    public init() {
        word = ""
        state = .None
    }

    mutating public func start() {
        state = .Started
    }

    mutating public func finish(string: String) {
        state = .Finished
        word = string
    }

    public mutating func abort() {
        state = .Aborted
        word = ""
    }

    public mutating func clear () {
        state = .None
        word = ""
    }

    // MARK: - Bridge

    public func bridgedWord() -> String {
        word
    }

    public func bridgedState() -> SKKRegistrationState {
        state
    }
}
